### PR TITLE
[CMake] Generate cmake target that other libraries can use

### PR DESCRIPTION
**Pull request by Mike-Devel**
... to express their dependency on this library and retrieve any
configuration information such as the include directory, binary
to link to (if any), transitive dependencies, necessary compiler
options or the required c++ standards level.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+cmake_minimum_required(VERSION 3.5)
+project(BoostThrowException LANGUAGES CXX)
+
+add_library(boost_throw_exception INTERFACE)
+add_library(Boost::throw_exception ALIAS boost_throw_exception)
+
+target_include_directories(boost_throw_exception INTERFACE include)
+
+target_link_libraries(boost_static_assert
```
Comment (pdimov) [inline CMakeLists.txt]: Wrong library name.

Comment (Mike-Devel) [reply CMakeLists.txt]: Urgs. Fixed

**Comment by pdimov** [comment]
What's the end goal? Towards which library are we advancing? :-)

**Comment by Mike-Devel** [comment]
> What's the end goal? Towards which library are we advancing? :-)

In my dream world? All
Personally, I'm currently interested in signals2, program_options, mp11, serialization, beast, circular_buffer and multiindex, but lets see, who in the dependency chain is willing to accept the cmake files. 

For now I'm just  picking some libraries from the lowest levels of `boostdep --module-levels` and try to gather feedback from you (the library maitainers). When config and the superporject have a proper cmake file, I'll have to put aside some time to go into "mass production" of PRs.

Thanks for adding a CMake file to boost.core by the way.

**Comment by glenfe** [comment]
@glenfe did a few already, but will do more when momentum picks up:
* https://github.com/boostorg/circular_buffer/blob/develop/CMakeLists.txt
* https://github.com/boostorg/concept_check/blob/develop/CMakeLists.txt
* https://github.com/boostorg/align/blob/develop/CMakeLists.txt

**Comment by Mike-Devel** [comment]
> But without Config merged, we're stuck. :-)

Yeah that worries me too. It is of course possible to work around that in various ways, but in the end it is questionable if that is a healthy way forward.

> The boost_test function from Assert should probably go someplace common such as libs/mincmake but I'm not sure if we have a mandate from heaven for that yet.

No I'm afraid not ;)

On a more serious note though: I don't want to give the impression to implement a cmake solution behind the back of the community (or parts of it) without a proper review process,. This is why I stick to those absolute minimal PRs that hopefully don't get in the way of whatever full-fledged cmake solution the boost community will finally settle on - whenever that happens. For the same reason I myself will not try to put utility functions into a common directory and will also keep the PR for the top level cmake file completely stupid.

Anyway, @glenfe, @pdimov: Thank you both very much for picking this up.

**Comment by pdimov** [comment]
In the meantime, there's https://github.com/pdimov/mincmake, used like in https://github.com/boostorg/assert/commit/a45445344357be8b5457397bd4086a097d0309c5.

**Comment by Mike-Devel** [comment]
Nice.

I should mention that winapi will not accept cmake related PRs until a full fleged, reviewed and accepted solution is available.

**Comment by glenfe** [comment]
> No I'm afraid not ;)

The mandate from heaven might require you to elicit a response from the Members listed [here](https://sites.google.com/a/boost.org/steering/).

**Comment by Mike-Devel** [comment]
@glenfe: Considering the shitstorm that happened the last time the steering committee endorsed cmake, people would probably lynch them if they officially endorsed this ;)

**Comment by pdimov** [comment]
FWIW, @steveire recommends we use lowercase `boost::` as the target namespace, to avoid confusion with the FindBoost targets and make the choice explicit. This may indeed save us headaches in the long run, and the time to decide is now.

**Comment by pdimov** [comment]
It's not possible to switch because FindBoost doesn't have imported targets for header-only libraries.

**Comment by MikeGitb** [comment]
Well, not yet for individual libraries, but it does provide `Boost::boost`

**Comment by pdimov** [comment]
Well the point is that projects using FindBoost will not work with our scheme anyway, and the idea of not using their convention is to make the choice of the scheme in use explicit and visible. If you see Boost::filesystem, the project uses FindBoost, if boost::filesystem - the new thing.

**Comment by Mike-Devel** [comment]
> Well the point is that projects using FindBoost will not work with our scheme anyway

It actually did work for me in the past. While it isn't ideal, it is not really a problem for a user to depend on an umbrella target for all the header only libraries (`Boost::boost`) and defining such an umbrella target is pretty straight forward. Also, I'd very much hope, that once  more and more boost header only libraries provide cmake targets, cmake itself will also forward those targets for consumation by the user. Of course if `boost` is whats used by likely candidates for a future, full fleged boost solution, then it makes sense to stay consistent with those solutions, althoug - as I said before - I'd personally prefer a more obvious difference between the two naming schemes rather than just a lower vs upper case.

In any case - as long as it isn't signed of by the ML (which I'm doubtfull will happen anytime soon) - I believe any naming scheme will be subject to change in the future anyway.

Again, I've not a strong opinion, which way to go, as long as I get something and in the end it is up to you library maintainers anyway what you are willing to accept into your libraries.

**Pull request merged**